### ISSUE TITLE
App reload breaks form submittal for forms that were loaded in browser before the reload

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#165]
**Reported by @frederikhabils on 24 Mar 2003 23:46 UTC**
When a makumba app (like karamba) is reloaded, then the users who were busy 
filling in a form, can not submit this form:

So the sequence is:
1/ user surfs to the forms, starts filling it in
2/ app is reloaded on the server 
3/ user want to submit

-->
Makumba invalid value error
Form responder cannot be found, probably due to server restart. Please reload 
the form page.
<--

It would be good to have this fixed. Otherwise, many users can be affected by 
the frequent reloads (necessary for BL fixes, etc).

Migrated-From: http://trac.makumba.org/ticket/363

**Comment by ghost** [comment]
**Comment by @stefanb on 25 Mar 2003 08:20 UTC**
This is expected behaviour.
If a developer changed the logic and did a reload, he probably wants new logic
to start functioning, not some old responder to be lurking around, handling form
submitals.

btw: Reload is much like complete restart (well, in theory at least :), except
that it is a lot faster.

But his can be annoying to users, so production reloads should not be so common,
preferably done at night with not many users filling in the forms.

Is there a way to initiate a new responder, eg by keeping a symbolic form name
("/lbg/lbgEdit.jsp#form1") in the session, apart from the numerical handler id
in the form itself?

that way a responder could probably be recreated somehow if the app was reloaded
and the session is still active (presently it stays active 30 minutes after last
 request). Numerical reference to the form handler in the form itself would
still take care of finding the form handler if user's session expired (in case
of long er form filling)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 25 Mar 2003 11:49 UTC**
sure this is "expected" behavior as things stand now, but that doesn't make it
impossible to improve. don't tend to be grounded in the present state of affairs :)

it should be possible to improve the user experience somehow. even if logic
changes, the form can respond with the new logic. what cannot change is the
structure of the form that is being filled in and responded to. that can be
serialized (saved to disk) somehow, instead of just keeping a hashtable cache
which dies at reload. saving it in the session is one way, though that means
that there will be a disk write for every session that executes that form, and
then the "size on disk" of the form responder becomes important.

in short, serailizing form responders is i think key to solving this. with the
new form response code isolated in org.makumba.controller (0.5.7), this should
be a bit easier to achieve.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 22 May 2003 20:41 UTC**
i think that if (after fixing bug 447) the structure of the original form is
known (or can be retreived) and the names of the <mak:input >s that did take
effect in the form are transmitted as a hidden argument (maybe some do not take
effect due to <%if %>s) then we do not need form responders anymore, so this can
get fixed that way

however, form responders are currently separated from the taglib, so fixing bug
5 will not imply eliminating form reponders, at least not in the first phase

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 10 May 2005 23:20 UTC**
During the 2nd Makumba Congress in Stockholm, the following design decision was
made:
One creation of a _new_ form responder, we check if it is already written to the
disk, if not, we write it to a file using standard Java object serialisation.
When looking for a responder, given its ID, if we do not find the responder in
the cache (due to a server restart), we try to read from the previously saved
file and restore the responder and therefore bypass the "Form responder cannot
be found" error.
Only in the case that the responder cannot be restored from the disk (most
probably due to deleting the files in the working directory), the old error appears.

The implementation is completed, but awaits testing before the bug can be set to
resolved.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 18 Jan 2006 04:05 UTC**
tested
